File: apps/backend/java-services/business-services/reservation-engine/src/main/java/com/modernreservation/reservationengine/entity/Tenant.java
package com.modernreservation.reservationengine.entity;

import com.modernreservation.reservationengine.enums.TenantType;
import com.modernreservation.reservationengine.enums.TenantStatus;
import jakarta.persistence.*;
import jakarta.validation.constraints.Email;
import jakarta.validation.constraints.NotNull;
import jakarta.validation.constraints.Pattern;
import lombok.*;
import org.hibernate.annotations.JdbcTypeCode;
import org.hibernate.type.SqlTypes;
import org.springframework.data.annotation.CreatedDate;
import org.springframework.data.annotation.LastModifiedDate;
import org.springframework.data.jpa.domain.support.AuditingEntityListener;

import java.time.LocalDateTime;
import java.util.Map;
import java.util.UUID;

/**
 * Tenant Entity - Multi-tenancy Organization
 *
 * Represents a tenant organization such as a hotel chain, franchise,
 * independent property, or property management company.
 */
@Entity
@Table(name = "tenants")
@EntityListeners(AuditingEntityListener.class)
@Getter
@Setter
@NoArgsConstructor
@AllArgsConstructor
@Builder
@ToString
@EqualsAndHashCode(onlyExplicitlyIncluded = true)
public class Tenant {

    @Id
    @GeneratedValue(strategy = GenerationType.AUTO)
    @Column(name = "id")
    @EqualsAndHashCode.Include
    private UUID id;

    @NotNull
    @Column(name = "name", nullable = false, length = 200)
    private String name;

    @NotNull
    @Pattern(regexp = "^[a-z0-9-]+$", message = "Slug must contain only lowercase letters, numbers, and hyphens")
    @Column(name = "slug", unique = true, nullable = false, length = 200)
    @EqualsAndHashCode.Include
    private String slug;

    @NotNull
    @Enumerated(EnumType.STRING)
    @Column(name = "type", nullable = false)
    private TenantType type;

    @NotNull
    @Enumerated(EnumType.STRING)
    @Column(name = "status", nullable = false)
    @Builder.Default
    private TenantStatus status = TenantStatus.TRIAL;

    // Contact Information
    @Email
    @NotNull
    @Column(name = "email", nullable = false, length = 255)
    private String email;

    @Column(name = "phone", length = 20)
    private String phone;

    @Column(name = "website", length = 500)
    private String website;

    // Address
    @Column(name = "address_line1", length = 255)
    private String addressLine1;

    @Column(name = "address_line2", length = 255)
    private String addressLine2;

    @Column(name = "city", length = 100)
    private String city;

    @Column(name = "state", length = 100)
    private String state;

    @Column(name = "postal_code", length = 20)
    private String postalCode;

    @Column(name = "country", length = 2)
    private String country; // ISO 3166-1 alpha-2

    // Business Information
    @Column(name = "tax_id", length = 100)
    private String taxId;

    @Column(name = "business_license", length = 100)
    private String businessLicense;

    @Column(name = "registration_number", length = 100)
    private String registrationNumber;

    // Configuration stored as JSONB
    @JdbcTypeCode(SqlTypes.JSON)
    @Column(name = "config", columnDefinition = "jsonb")
    private Map<String, Object> config;

    // Subscription stored as JSONB
    @JdbcTypeCode(SqlTypes.JSON)
    @Column(name = "subscription", columnDefinition = "jsonb")
    private Map<String, Object> subscription;

    // Metadata stored as JSONB
    @JdbcTypeCode(SqlTypes.JSON)
    @Column(name = "metadata", columnDefinition = "jsonb")
    private Map<String, Object> metadata;

    // Audit Fields
    @CreatedDate
    @Column(name = "created_at", nullable = false, updatable = false)
    private LocalDateTime createdAt;

    @LastModifiedDate
    @Column(name = "updated_at")
    private LocalDateTime updatedAt;

    @Column(name = "created_by", length = 100)
    private String createdBy;

    @Column(name = "updated_by", length = 100)
    private String updatedBy;

    // Soft Delete
    @Column(name = "deleted_at")
    private LocalDateTime deletedAt;

    @Column(name = "deleted_by", length = 100)
    private String deletedBy;

    // Optimistic Locking
    @Version
    @Column(name = "version")
    private Long version;

    /**
     * Check if tenant is soft deleted
     */
    public boolean isDeleted() {
        return deletedAt != null;
    }

    /**
     * Check if tenant is active
     */
    public boolean isActive() {
        return status == TenantStatus.ACTIVE && !isDeleted();
    }

    /**
     * Soft delete the tenant
     */
    public void softDelete(String deletedBy) {
        this.deletedAt = LocalDateTime.now();
        this.deletedBy = deletedBy;
        this.status = TenantStatus.CANCELLED;
    }
}

File: libs/shared/schemas/src/entities/domains/tenant.ts
/**
 * Tenant Domain Schemas
 * Multi-tenancy support for hotel chains, franchises, and independent properties
 */

import { z } from "zod";
import { AuditFieldsSchema, SoftDeleteFieldsSchema, UUIDSchema } from "./common";

// =============================================================================
// ENUMS
// =============================================================================

/**
 * Tenant type classification
 */
export const TenantTypeSchema = z.enum([
  "CHAIN", // Hotel chain (e.g., Marriott, Hilton)
  "INDEPENDENT", // Independent hotel/property
  "FRANCHISE", // Franchised property
  "MANAGEMENT_COMPANY", // Property management company
  "VACATION_RENTAL", // Vacation rental company (Airbnb host, etc.)
]);

export type TenantType = z.infer<typeof TenantTypeSchema>;

/**
 * Subscription plan tiers
 */
export const SubscriptionPlanSchema = z.enum([
  "FREE", // Free tier (limited properties)
  "STARTER", // Small properties (1-5 properties)
  "PROFESSIONAL", // Medium properties (6-20 properties)
  "ENTERPRISE", // Large chains (unlimited properties)
  "CUSTOM", // Custom enterprise plan
]);

export type SubscriptionPlan = z.infer<typeof SubscriptionPlanSchema>;

/**
 * Tenant status
 */
export const TenantStatusSchema = z.enum([
  "ACTIVE", // Active and operational
  "SUSPENDED", // Temporarily suspended
  "TRIAL", // Trial period
  "EXPIRED", // Subscription expired
  "CANCELLED", // Account cancelled
]);

export type TenantStatus = z.infer<typeof TenantStatusSchema>;

// =============================================================================
// TENANT CONFIGURATION
// =============================================================================

/**
 * Tenant configuration and settings
 */
export const TenantConfigSchema = z.object({
  // Branding
  brandingEnabled: z.boolean().default(true),
  logoUrl: z.string().url().optional(),
  primaryColor: z.string().regex(/^#[0-9A-Fa-f]{6}$/).optional(),
  secondaryColor: z.string().regex(/^#[0-9A-Fa-f]{6}$/).optional(),

  // Features
  enableMultiProperty: z.boolean().default(true),
  enableChannelManager: z.boolean().default(false),
  enableAdvancedReporting: z.boolean().default(false),
  enablePaymentProcessing: z.boolean().default(true),
  enableLoyaltyProgram: z.boolean().default(false),

  // Limits
  maxProperties: z.number().int().positive().default(5),
  maxUsers: z.number().int().positive().default(10),
  maxReservationsPerMonth: z.number().int().positive().optional(),

  // Localization
  defaultCurrency: z.string().length(3).default("USD"),
  defaultLanguage: z.string().length(2).default("en"),
  defaultTimezone: z.string().default("America/New_York"),

  // Notifications
  enableEmailNotifications: z.boolean().default(true),
  enableSmsNotifications: z.boolean().default(false),
  enablePushNotifications: z.boolean().default(true),
});

export type TenantConfig = z.infer<typeof TenantConfigSchema>;

// =============================================================================
// SUBSCRIPTION
// =============================================================================

/**
 * Tenant subscription details
 */
export const TenantSubscriptionSchema = z.object({
  plan: SubscriptionPlanSchema,
  startDate: z.coerce.date(),
  endDate: z.coerce.date().optional(),
  trialEndsAt: z.coerce.date().optional(),
  autoRenew: z.boolean().default(true),
  paymentMethod: z.string().optional(), // Stripe payment method ID

  // Billing
  billingCycle: z.enum(["MONTHLY", "YEARLY", "CUSTOM"]).default("MONTHLY"),
  amount: z.number().nonnegative(),
  currency: z.string().length(3).default("USD"),
  nextBillingDate: z.coerce.date().optional(),

  // Status
  isTrial: z.boolean().default(false),
  isActive: z.boolean().default(true),
  cancelledAt: z.coerce.date().optional(),
});

export type TenantSubscription = z.infer<typeof TenantSubscriptionSchema>;

// =============================================================================
// MAIN TENANT SCHEMA
// =============================================================================

/**
 * Tenant (Organization) Schema
 * Represents a hotel chain, franchise, or independent property owner
 */
export const TenantSchema = z.object({
  id: UUIDSchema,

  // Basic Info
  name: z.string().min(1).max(200),
  slug: z.string().regex(/^[a-z0-9-]+$/), // URL-friendly identifier
  type: TenantTypeSchema,
  status: TenantStatusSchema.default("ACTIVE"),

  // Contact Info
  email: z.string().email(),
  phone: z.string().optional(),
  website: z.string().url().optional(),

  // Address
  addressLine1: z.string().optional(),
  addressLine2: z.string().optional(),
  city: z.string().optional(),
  state: z.string().optional(),
  postalCode: z.string().optional(),
  country: z.string().length(2).optional(), // ISO 3166-1 alpha-2

  // Business Info
  taxId: z.string().optional(), // Tax identification number
  businessLicense: z.string().optional(),
  registrationNumber: z.string().optional(),

  // Configuration
  config: TenantConfigSchema.default({}),

  // Subscription
  subscription: TenantSubscriptionSchema,

  // Metadata
  metadata: z.record(z.string(), z.any()).optional(), // Custom fields

  // Timestamps & Audit
  ...AuditFieldsSchema.shape,
  ...SoftDeleteFieldsSchema.shape,
});

export type Tenant = z.infer<typeof TenantSchema>;

// =============================================================================
// TENANT CONTEXT
// =============================================================================

/**
 * Tenant context for requests
 * Used in GraphQL context and API requests
 */
export const TenantContextSchema = z.object({
  tenantId: UUIDSchema,
  tenantName: z.string(),
  tenantType: TenantTypeSchema,
  subscriptionPlan: SubscriptionPlanSchema,
  isActive: z.boolean(),
  features: z.object({
    multiProperty: z.boolean(),
    channelManager: z.boolean(),
    advancedReporting: z.boolean(),
    paymentProcessing: z.boolean(),
    loyaltyProgram: z.boolean(),
  }),
});

export type TenantContext = z.infer<typeof TenantContextSchema>;

// =============================================================================
// TENANT CREATION & UPDATE
// =============================================================================

/**
 * Schema for creating a new tenant
 */
export const CreateTenantSchema = TenantSchema.omit({
  id: true,
  createdAt: true,
  updatedAt: true,
  createdBy: true,
  updatedBy: true,
  deletedAt: true,
  deletedBy: true,
}).extend({
  // Override to make fields required for creation
  name: z.string().min(2).max(200),
  email: z.string().email(),
  type: TenantTypeSchema,
});

export type CreateTenant = z.infer<typeof CreateTenantSchema>;

/**
 * Schema for updating tenant
 */
export const UpdateTenantSchema = TenantSchema.partial().omit({
  id: true,
  createdAt: true,
  createdBy: true,
  deletedAt: true,
  deletedBy: true,
});

export type UpdateTenant = z.infer<typeof UpdateTenantSchema>;

// =============================================================================
// TENANT FILTERS
// =============================================================================

/**
 * Filter schema for querying tenants
 */
export const TenantFilterSchema = z.object({
  id: UUIDSchema.optional(),
  slug: z.string().optional(),
  type: TenantTypeSchema.optional(),
  status: TenantStatusSchema.optional(),
  subscriptionPlan: SubscriptionPlanSchema.optional(),
  search: z.string().optional(), // Search by name, email, etc.
  isActive: z.boolean().optional(),
  includeDeleted: z.boolean().default(false),
});

export type TenantFilter = z.infer<typeof TenantFilterSchema>;
